### PR TITLE
Simplify KMeans code

### PR DESCRIPTION
**Pull request by Komzpa**
https://trac.osgeo.org/postgis/ticket/3977

### CODE REVIEW

File: liblwgeom/lwkmeans.c [hunk 1]
```diff
@@ -1,77 +1,195 @@
...
-	POINT2D sum;
-	POINT2D **pts = (POINT2D**)objs;
-	POINT2D *center = (POINT2D*)centroid;
-
-	sum.x = sum.y = 0.0;
+	unsigned int i;
+	unsigned int* weights;
 
-	if (num_objs <= 0) return;
+	weights = lwalloc(sizeof(int) * config->k);
```
Comment (pramsey) [inline liblwgeom/lwkmeans.c]: Consider moving this to config, so you don't have to alloc/free it on each iteration.

Comment (Komzpa) [reply liblwgeom/lwkmeans.c]: Moved out of function. Got rid of config altogether, -50 lines.

**Comment by Komzpa** [comment]
@pramsey @tomvantilburg please take a look and wave an OK if it's fine to commit.

**Comment by tomvantilburg** [comment]
I'm happy with the new implementation. It is certainly faster, and much better than what we had :wave:
Just for the looks of it, here's the result of the clustering on a regular grid (bluer is larger maxdistance, from 1.4 to 4.5):
image

**Comment by tomvantilburg** [comment]
And the situation as it was previously (curiously, dropping 2 clusters out of 1000):
image

**Comment by pramsey** [comment]
:wave:

**Pull request closed without merging**